core/src/main/java/org/elasticsearch/index/mapper/UidFieldMapper.java

```diff
@@ -96,20 +114,81 @@ public String typeName() {
+                return new IndexFieldData.Builder() {
+                    @Override
+                    public IndexFieldData<?> build(IndexSettings indexSettings, MappedFieldType fieldType, IndexFieldDataCache cache,
+                            CircuitBreakerService breakerService, MapperService mapperService) {
+                        MappedFieldType idFieldType = mapperService.fullName(IdFieldMapper.NAME);
+                        IndexOrdinalsFieldData idFieldData = (IndexOrdinalsFieldData) idFieldType.fielddataBuilder()
+                                .build(indexSettings, idFieldType, cache, breakerService, mapperService);
+                        final String type = mapperService.types().iterator().next();
+                        return new UidIndexFieldData(indexSettings.getIndex(), type, idFieldData);
+                    }
+                };
+            } else {
+                // Old index, _uid was indexed
+                return new PagedBytesIndexFieldData.Builder(
+                        TextFieldMapper.Defaults.FIELDDATA_MIN_FREQUENCY,
+                        TextFieldMapper.Defaults.FIELDDATA_MAX_FREQUENCY,
+                        TextFieldMapper.Defaults.FIELDDATA_MIN_SEGMENT_SIZE);
+            }
+        }
+
+        @Override
+        public Query termQuery(Object value, @Nullable QueryShardContext context) {
+            return termsQuery(Arrays.asList(value), context);
+        }
+
+        @Override
+        public Query termsQuery(List<?> values, @Nullable QueryShardContext context) {
+            if (indexOptions() != IndexOptions.NONE) {
+                return super.termsQuery(values, context);
+            }
+            Collection<String> indexTypes = context.getMapperService().types();
+            if (indexTypes.isEmpty()) {
+                return new MatchNoDocsQuery("No types");
+            }
+            assert indexTypes.size() == 1;
+            BytesRef indexType = indexedValueForSearch(indexTypes.iterator().next());
+            BytesRefBuilder prefixBuilder = new BytesRefBuilder();
+            prefixBuilder.append(indexType);
+            prefixBuilder.append((byte) '#');
+            BytesRef expectedPrefix = prefixBuilder.get();
+            List<BytesRef> ids = new ArrayList<>();
+            for (Object uid : values) {
+                BytesRef uidBytes = indexedValueForSearch(uid);
+                if (StringHelper.startsWith(uidBytes, expectedPrefix)) {
+                    BytesRef id = new BytesRef();
+                    id.bytes = uidBytes.bytes;
+                    id.offset = uidBytes.offset + expectedPrefix.length;
+                    id.length = uidBytes.length - expectedPrefix.length;
+                    ids.add(id);
+                }
+            }
+            return new TermInSetQuery(IdFieldMapper.NAME, ids);
+        }
+    }
+
+    static MappedFieldType defaultFieldType(Settings indexSettings) {
```

Comment:
same what I mentioned above, we have `IndexSettings` available here can we maybe pass a boolean down? also  maybe we should just reference a static method from `IdFieldMapper` here?